File: 121.Best-Time-to-Buy-and-Sell-Stock/dp.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        max_profit, min_so_far = 0, prices[0]

        # O(n)
        for i in range(1, len(prices) + 1):
            max_profit = max(max_profit, prices[i - 1] - min_so_far)
            min_so_far = min(min_so_far, prices[i - 1])

        return max_profit

File: 121.Best-Time-to-Buy-and-Sell-Stock/dp.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut min_price, mut max_profit) = (prices[0], 0);

        // O(n)
        for price in prices {
            max_profit = max_profit.max(price - min_price);
            min_price = min_price.min(price);
        }

        max_profit
    }
}
